storage/src/data_access_layer/balance.rs

```diff
@@ -305,15 +403,149 @@ impl BalanceRequest {
     pub fn proof_handling(&self) -> ProofHandling {
         self.proof_handling
     }
+
+    /// Returns gas hold balance handling.
+    pub fn gas_hold_balance_handling(&self) -> GasHoldBalanceHandling {
+        self.gas_hold_balance_handling
+    }
+}
+
+pub trait AvailableBalanceChecker {
+    /// Calculate and return available balance.
+    #[allow(clippy::result_unit_err)]
+    fn available_balance(
+        &self,
+        block_time: BlockTime,
+        total_balance: U512,
+        handling: HoldBalanceHandling,
+        interval: TimeDiff,
+    ) -> Result<U512, BalanceError> {
+        if self.is_empty() {
+            return Ok(total_balance);
+        }
+        let held = match handling {
+            HoldBalanceHandling::Accrued => self.accrued(),
+            HoldBalanceHandling::Amortized => {
+                let mut held = U512::zero();
+                let block_time = block_time.value();
+                let interval = interval.millis();
+
+                for (hold_created_time, holds) in self.holds() {
```

Comment:
I don't think this is entirely correct. If the map holds both gas holds and processing holds then the amortization is applied for both of them. I don't think processing holds should be ever amortized.
I think `Self::holds()` should probably be called `Self::amortizable_holds()` and the implementations of that should return only holds that can be amortized. In our case it would return only gas holds, without processing holds.

Reply:
Ah, yes. Good catch.

Reply:
ok, tweaked the logic to split out handling by hold kind